### DISCUSSION TITLE
Progressive enhancement or graceful degradation?

### DISCUSSION THREAD

**Discussion by LukaszJaro** [filamentphp/filament#14235] [Help] [answered]
### Package

Other

### Package Version

v3

### How can we help you?

When JavaScript is disabled in Filament, the page just returns a white page with no content.

Progressive enhancement is a design philosophy that provides a baseline of essential content and functionality to as many users as possible, while delivering the best possible experience only to users of the most modern browsers that can run all the required code.
- https://www.gov.uk/service-manual/technology/using-progressive-enhancement

There is many scenarios where JS fails, preventing users from completing their workflow. How is this best handled using filament? Is it possible to create a fall back?

**Comment by LukaszJaro** [comment]
This partially answers the question, since Filament is dependent on livewire, but still wondering how Devs handle this.

**Livewire is inefficient**
This doesn't matter so much in 2020, where bandwidth is mostly fast and cheap. However, the reality of the situation is that Livewire makes AJAX Requests for every. single. thing. that you do.

It also makes Livewire the polar opposite of a [Progressive Web App](https://developers.google.com/web/progressive-web-apps). If you lose connection, the most basic functions of your app’s presentation layer will cease to function. Modals won’t open, dropdowns won’t… drop down, et cetera. In Livewire’s defense here, it does provide functionality that helps with this. It makes it easy to alert the user that the app will be malfunctioning, but it will still malfunciton.

It also replaces entire chunks of HTML at a time, and forces the browser to re-draw it. Depending on how large your Livewire component is, this could actually make a noticeable difference to the user. In contrast, frameworks like Vue and React use a “[Virtual Dom](https://reactjs.org/docs/faq-internals.html)”, and a highly efficient diffing algorithm in order to make their updates. And they cut out the network call completely.
Source: https://medium.com/@nathanpisarski/making-the-case-against-laravel-livewire-a88c68ca0627

**Reply by mgkimsal** [reply, marked as answer]
So.. do not use it?  

Filament provides a crazy amount of productive value for some (many?) use cases, but nothing fits every need.  I've yet to find any toolkit that provides this level of productivity for form-oriented web applications in any language.  And part of its ability to deliver that is because of livewire - precisely because it lets you avoid having to write much javascript (or often, not having to write any at all).  It comes with some constraints.  

Perhaps someday someone will write a 'filament for vue', and provide offline capabilities, data syncing, conflict resolution, and all that, while allowing you to write declarative PHP.  Until that time, I'll be using filament for many projects.

To answer the initial question, no, there's not a way to provide a fallback.

**Reply by LukaszJaro** [reply]
I'm well aware of how powerful Filament is, which is why I wanted to use it on more projects but some projects have strict requirements on progressive enhancements so alternative approach would be needed. Thanks for your answer.
